File: lbr_demos/lbr_demos_fri_ros2_cpp/src/torque_sine_overlay_node.cpp
#include <functional>
#include <math.h>
#include <memory>
#include <string>

// include fri for session state
#include "friClientIf.h"

#include "rclcpp/rclcpp.hpp"

// include lbr_fri_msgs
#include "lbr_fri_msgs/msg/lbr_state.hpp"
#include "lbr_fri_msgs/msg/lbr_torque_command.hpp"

class TorqueSineOverlayNode : public rclcpp::Node {
  static constexpr double amplitude_ = 15.;  // Nm
  static constexpr double frequency_ = 0.25; // Hz

public:
  TorqueSineOverlayNode(const std::string &node_name) : Node(node_name), phase_(0.) {
    // create publisher to /lbr/command/torque
    lbr_torque_command_pub_ =
        this->create_publisher<lbr_fri_msgs::msg::LBRTorqueCommand>("/lbr/command/torque", 1);

    // create subscription to /lbr/state
    lbr_state_sub_ = this->create_subscription<lbr_fri_msgs::msg::LBRState>(
        "/lbr/state", 1,
        std::bind(&TorqueSineOverlayNode::on_lbr_state_, this, std::placeholders::_1));
  };

protected:
  void on_lbr_state_(const lbr_fri_msgs::msg::LBRState::SharedPtr lbr_state) {
    lbr_torque_command_.joint_position = lbr_state->ipo_joint_position;

    if (lbr_state->session_state == KUKA::FRI::COMMANDING_ACTIVE) {
      // overlay torque sine wave on 4th joint
      lbr_torque_command_.torque[3] = amplitude_ * sin(phase_);
      phase_ += 2 * M_PI * frequency_ * lbr_state->sample_time;

      lbr_torque_command_pub_->publish(lbr_torque_command_);
    } else {
      // reset phase
      phase_ = 0.;
    }
  };

  double phase_;

  rclcpp::Publisher<lbr_fri_msgs::msg::LBRTorqueCommand>::SharedPtr lbr_torque_command_pub_;
  rclcpp::Subscription<lbr_fri_msgs::msg::LBRState>::SharedPtr lbr_state_sub_;

  lbr_fri_msgs::msg::LBRTorqueCommand lbr_torque_command_;
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<TorqueSineOverlayNode>("torque_sine_overlay_node"));
  rclcpp::shutdown();
  return 0;
};

File: lbr_demos/lbr_demos_py/lbr_demos_py/torque_sine_overlay_node.py
import math

import rclpy
from rclpy.node import Node

# import lbr_fri_msgs
from lbr_fri_msgs.msg import LBRState, LBRTorqueCommand


class TorqueSineOverlayNode(Node):
    def __init__(self, node_name: str) -> None:
        super().__init__(node_name)

        self._amplitude = 15.0  # Nm
        self._frequency = 0.25  # Hz
        self._phase = 0.0
        self._lbr_torque_command = LBRTorqueCommand()

        # create publisher to /lbr/command/torque
        self._lbr_torque_command_pub = self.create_publisher(
            LBRTorqueCommand, "/lbr/command/torque", 1
        )

        # create subscription to /lbr_state
        self._lbr_state_sub = self.create_subscription(
            LBRState, "/lbr/state", self._on_lbr_state, 1
        )

    def _on_lbr_state(self, lbr_state: LBRState) -> None:
        self._lbr_torque_command.joint_position = lbr_state.ipo_joint_position

        if lbr_state.session_state == 4:  # KUKA::FRI::COMMANDING_ACTIVE == 4
            # overlay torque sine wave on 4th joint
            self._lbr_torque_command.torque[3] = self._amplitude * math.sin(self._phase)
            self._phase += 2 * math.pi * self._frequency * lbr_state.sample_time

            self._lbr_torque_command_pub.publish(self._lbr_torque_command)
        else:
            # reset phase
            self._phase = 0.0


def main(args: list = None) -> None:
    rclpy.init(args=args)
    rclpy.spin(TorqueSineOverlayNode("torque_sine_overlay_node"))
    rclpy.shutdown()
